src/odemis/odemisd/modelgen.py

```diff
@@ -277,6 +278,27 @@ def __init__(self, inst_file, settings_file=None, container=None, create_sub_con
 
         # TODO: check there is no cyclic dependencies on the parents/children
 
+    def _check_structure(self):
+        """
+        The AST should be a dict[str, dict[str, ...]]: name of component -> definition
+        Each component definition must have at least the "role" key
+        """
+        if not isinstance(self.ast, dict):
+            raise ParseError(f"Microscope file must be a dictionary of component names -> definition, but got a {type(self.ast).__name__}")
+
+        for n, a in self.ast.items():
+            if not isinstance(n, str):
+                raise ParseError(f"Component name should be string but got: {n}")
+            if not isinstance(a, dict):
+                raise ParseError(f"Definition of component \"{n}\" should be a dict but got a {type(a).__name__}")
+
+            # Check inside the component definition
+            for arg in a.keys():
+                if not isinstance(arg, str):
+                    raise ParseError(f"Definition of component \"{n}\" should only have keys of type str but got a {type(arg).__name__}")
+            if not "role" in a:
+                raise SemanticError(f"Definition of component \"{n}\" is missing a key 'role'.")
```

Comment:
You could point out that 'role' is a mandatory/required key.
```suggestion
                raise SemanticError(f"Definition of component \"{n}\" is missing the mandatory key 'role'.")
```